File: php-sdk/rabbit.0.9.1.php
<?php
namespace Bad;
/**
 * author 逆雪寒
 * version 0.9.1
 */
class Rabbit {

	const YES = 1;
	const NO = 0;

	private static $instance = NULL;
	private $host = '';

	public $trace = self::NO;

	private function __construct($host,$port){ 
		$this->host($host,$port);
	}
	 
	public function __clone(){
		trigger_error('Clone is not allow!',E_USER_ERROR);
	}

	/**
	 * 调试接口返回.
	 *
	 * @param string $msg .
	 * @return $this A reference to the current instance.
	 */
	private function trace($msg) {
		if($this->trace){
			exit($msg);
		}
	}

	private function decode($data) {
		$this->trace($data);
		return json_decode($data,true);
	}

	private function host($host,$port) {
		$this->host = "http://" . $host . ":" . $port."/";
	}

	public function factory($host = 'localhost',$port = 9394) {
		if(is_null(self::$instance)) {
			self::$instance = new self($host,$port);
		}
		return self::$instance;
	}

	private function request($do,Array $parameter = [],$method = 'GET') {
		$query = '';

		$ch = curl_init(); 

		if($method != 'GET') {
			curl_setopt($ch, CURLOPT_POSTFIELDS, $parameter);
			curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);

		}else{
			$query = "?" . http_build_query($parameter);
		}

		curl_setopt($ch, CURLOPT_URL, $this->host.$do.$query);
		curl_setopt($ch, CURLOPT_TIMEOUT,60);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER,true);

		$data = curl_exec($ch);
		if(curl_errno($ch)){ 
		  return false;
		}
		curl_close($ch);
		return $data;
	}

	/**
	 * 内容过滤.
	 *
	 * @param string 	$contents 要过滤的内容
	 * @return bool or array
	 */
	public function filter($contents) {
		$result = $this->request("filter",['contents' => $contents],'POST');
		if($result !== false) {
			$code = $this->decode($result);
			if(isset($code['success']) && $code['success'] === self::NO) {
				return false;
			}
			return $code;
		}
		return false;
	}

	/**
	 * 色图过滤.
	 *
	 * @param string 	$file 要过滤的图片 本地地址 或者 网络地址
	 * @return bool or int  分数  建议：65 - 85判断为性感 85以上为色情
	 */
	public function porn($file) {
		$result = $this->request("porn",['file' => $file],'POST');
		if($result !== false) {
			$code = $this->decode($result);
			if(isset($code['success']) && $code['success'] === self::NO) {
				return false;
			}
			return $code['score'];
		}
		return false;
	}

	/**
	 * 脏词删除.
	 *
	 * @param int 	$id 脏词id （必须）
	 * @return bool or array
	 */
	public function delete($id) {
		if(!$id) return false;
		$result = $this->request("delete",['id' => $id],'DELETE');
		if($result !== false) {
			$code = $this->decode($result);
			if(isset($code['success']) && $code['success'] === self::NO) {
				return false;
			}
			return true;
		}
		return false;
	}

	/**
	 * 添加脏词.
	 *
	 * @param array 
	 * 			string  word 脏词 
	 * 			int 	category 脏词分类
	 * 			int 	rate 黑名单OR灰名单 1 or 2
 	 * 			int 	correct 是否畸形纠正  1  or 2
	 * @return bool or array.
	 */
	public function create(Array $parameter) {
		$result = $this->request("create",$parameter,'POST');
		if($result !== false) {
			$code = $this->decode($result);
			if(isset($code['success']) && $code['success'] === self::NO) {
				return false;
			}
			return true;
		}
		return false;
	}

	/**
	 * 修改脏词.
	 *
	 * @param array 
	 *			int 	id 脏词id （必须）
	 * 			string  word 脏词 
	 * 			int 	category 脏词分类
	 * 			int 	rate 黑名单OR灰名单 1 or 2
 	 * 			int 	correct 是否畸形纠正  1  or 2
	 * @return bool
	 */
	public function revise(Array $parameter) {

		if(!isset($parameter['id']) || count($parameter) < 1) {
			return false;
		}

		$result = $this->request("revise",$parameter,'PUT');
		if($result !== false) {
			$code = $this->decode($result);
			if(isset($code['success']) && $code['success'] === self::NO) {
				return false;
			}
			return true;
		}
		return false;
	}

	/**
	 * 查询脏词.
	 *
	 * @param array 
	 *			int 	id 脏词id
	 * 			string  word 脏词 
	 * 			int 	category 脏词分类
	 * 			int 	rate 黑名单OR灰名单 1 or 2
 	 * 			int 	correct 是否畸形纠正  1  or 2
	 * @return bool or array
	 */
	public function query(Array $parameter) {
		$result = $this->request("query",$parameter,'GET');
		if($result !== false) {
			$code = $this->decode($result);

			if(isset($code['success']) && $code['success'] === self::NO) {
				return false;
			}
			return $code;
		}
		return false;
	}

	/**
	 * 脏词分类.
	 * ID:分类名   1:个性化  2:低俗信息  3:灌水信息 5:政治敏感 6:违约广告  7:跨站追杀  8:色情信息 9:违法信息  10:垃圾广告
	 * @return bool or array
	 */
	public function category() {
		$result = $this->request("category");
		if($result !== false) {
			$code = $this->decode($result);

			if(isset($code['success']) && $code['success'] === self::NO) {
				return false;
			}
			
			return $code;
		}
		return false;
	}

	/**
	 * 脏词重新载入.
	 *
	 * @return bool
	 */
	public function reload() {
		$result = $this->request("reload");
		if($result !== false) {
			$code = $this->decode($result);

			if(isset($code['success']) && $code['success'] == self::NO) {
				return false;
			}
			return true;
		}
		return false;
	}
}

$rabbit = Rabbit::factory('127.0.0.1',9394);

// $rabbit->trace = true;

//$data = $rabbit->filter('万人敬仰+大兄弟 15009987776');

// $data = $rabbit->create([
// 			'word' => '什么狗屎+猪头三:phone',
// 			'category' => 3,
// 			'rate' => 1,
// 			'correct' => 1
// ]);

//$data = $rabbit->delete(29);

// $data = $rabbit->reload();

$score = $rabbit->porn("/Users/yehuiming/Downloads/6609c93d70cf3bc7e43db93dd500baa1cd112a25.jpg");

var_dump($score);

File: python-sdk/rabbit.0.9.1.py
# -*- coding: utf-8 -*-
"""
https://github.com/nixuehan/rabbit
"""
__author__ = '逆雪寒'
__version__ = '0.9.1'

try:
	import requests
except ImportError,msg:
	raise ImportError("Please install python requests. pip install requests")


class RabbitException(Exception):
	pass


class Rabbit(object):
	"""doc:  https://github.com/nixuehan/rabbit. """

	YES,NO = 1,0

	def __init__(self,host,port,timeout=2,trace=0):
		self.host,self.timeout,self.trace = "http://" + host + ":" + str(port) + "/",timeout,trace

	def __do(self,do):
		return self.host + do

	def __trace(self,msg):
		if self.trace == self.YES:
			print msg
		pass

	def filter(self,contents):
		"""
		过滤内容
		"""
		try:
			result = requests.post(self.__do('filter'),params={'contents':contents},timeout=self.timeout)
			data = result.json()
			if ("success" in data) and (data["success"] == self.NO):
				self.__trace(data)
				return False
			return data

		except Exception,msg:
			raise RabbitException(msg)


	def porn(self,filePath):
		"""
		过滤色图 score 建议：65 - 85判断为性感 85以上为色情
		"""
		try:
			result = requests.post(self.__do('porn'),params={'file':filePath},timeout=self.timeout)
			data = result.json()
			if ("success" in data) and (data["success"] == self.NO):
				self.__trace(data)
				return False
			return data['score']

		except Exception,msg:
			raise RabbitException(msg)


	def delete(self,id):
		"""
		删除脏词
		"""
		try:
			result = requests.delete(self.__do('delete'),params={'id':id},timeout=self.timeout)
			data = result.json()
			if ("success" in data) and (data["success"] == self.NO):
				return False
			return data.json()
		except Exception,msg:
			raise RabbitException(msg)

	def create(self,info):
		"""
		创建脏词

		@param dict
		string  word 脏词 
		int 	category 脏词分类
		int 	rate 黑名单OR灰名单 1 or 2
		int 	correct 是否畸形纠正  1  or 2
		"""
		try:
			result = requests.post(self.__do('create'),params=info,timeout=self.timeout)
			data = result.json()
			if ("success" in data) and (data["success"] == self.NO):
				self.__trace(data)
				return False
			return data

		except Exception,msg:
			raise RabbitException(msg)		

	def revise(self,info):
		"""
		修改脏词

		@param dict
		int 	id 脏词id
		string  word 脏词
		int 	category 脏词分类
		int 	rate 黑名单OR灰名单 1 or 2
		int 	correct 是否畸形纠正  1  or 2
		"""
		try:
			result = requests.put(self.__do('revise'),params=info,timeout=self.timeout)
			data = result.json()
			if ("success" in data) and (data["success"] == self.NO):
				self.__trace(data)
				return False
			else:
				return True
			return data

		except Exception,msg:
			raise RabbitException(msg)

	def query(self,info):
		"""
		查询脏词信息

		@param dict
		int 	id 脏词id
		string  word 脏词
		int 	category 脏词分类
		int 	rate 黑名单OR灰名单 1 or 2
		int 	correct 是否畸形纠正  1  or 2
		"""
		try:
			result = requests.get(self.__do('query'),params=info,timeout=self.timeout)
			data = result.json()
			if ("success" in data) and (data["success"] == self.NO):
				self.__trace(data)
				return False
			return data

		except Exception,msg:
			raise RabbitException(msg)

	def reload(self):
		"""
		脏词重载
		"""
		try:
			result = requests.get(self.__do('reload'),timeout=self.timeout)
			data = result.json()
			if ("success" in data) and (data["success"] == self.NO):
				return False
			return True
		except Exception,msg:
			raise RabbitException(msg)


rabbit = Rabbit(host='127.0.0.1',port=9394)

try:
	score = rabbit.porn("https://ss0.bdstatic.com/94oimg?image&quality=100&size=b4000_4000&sec=1465993408&di=db6c2337ac2ffe0cab3a6410e987098c&src=http://t1.mmonly.cc/uploads/allimg/20150727/h2pg440kruf.jpg")
except RabbitException,msg:
	score = False

print score
# print rabbit.query({'category':2,'rate':2})

# result = rabbit.filter("我吃蒙汗药")
# if(result['hit']):
# 	print 'hit'
# else:
# 	print 'no'